### PR TITLE
[#4551] One-click unsubscribe from Email Alerts

### LINKED ISSUE #4551: Should be able to unsubscribe from track emails without signing in
Currently there is an "Alter your subscription" link, which is a very high barrier to entry for an email subscription; you should not have to log in in order to stop getting an email.

### PR DESCRIPTION
**Pull request by garethrees**
## Relevant issue(s)

Fixes #4551

## What does this do?

Adds "one-click-unsubscribe" to `TrackMailer#event_digest` emails (sent by the alert-tracks daemon).

## Why was this needed?

Requiring users to sign in to unsubscribe from email alerts adds unnecessary effort, and encourages users to mark genuine emails as spam. This affects our reputation with email providers.

## Implementation notes

`User::EmailAlerts` adds advanced capability for managing a `User`'s email alerts, namely adding a token generator and verifier so that we can look up a `User` by a unique token and then change their email preferences.

Note that `User#receive_email_alerts` only manages notifications from `TrackMailer#event_digest`, so I've added messaging to that effect on the unsubscribe confirmation template.

#4603 covers adding additional notification management options.

## Screenshots

The unsubscribe link:

screen shot 2018-11-23 at 17 34 59

Successfully unsubscribing:

screen shot 2018-11-23 at 17 21 38

Passing an invalid token:

screen shot 2018-11-23 at 17 48 19

## Notes to reviewer

This was a quick way to setup and send a basic `event_digest` email:

```ruby
user = User.find(3)

track_thing = TrackThing.first

xapian_object = ActsAsXapian::Search.new([InfoRequestEvent], 'chicken',
                                                 :sort_by_prefix => 'described_at',
                                                 :sort_by_ascending => true,
                                                 :collapse_by_prefix => nil,
                                                 :limit => 100)


alert_results = []
alert_results.push(xapian_object.results.first)                                               
email_about_things = []
email_about_things.push([track_thing, alert_results, xapian_object])

TrackMailer.event_digest(user, email_about_things).deliver_now
```

This should work if you have the fixture data around, but I'm sure if you tweak the user ID / search term you can find something workable.

### CODE REVIEW

File: app/views/users/email_alerts/destroy.html.erb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<h1><%= _('Email Alerts Disabled') %></h1>
+
+<p>
+  <%= _('Email alerts for things you follow have now been disabled.') %>
+</p>
+
+<p>
+  <%= _('<a href="{{sign_in_link}}">Sign in</a> to get updates on things ' \
```
Comment (lizconlan) [inline app/views/users/email_alerts/destroy.html.erb]: Seeing this when signed in is a little odd, maybe a link to the wall instead for signed in users?

File: app/views/track_mailer/event_digest.text.erb [hunk 2]
```diff
@@ -75,4 +75,10 @@
 <% _('Please click on the link below to cancel or alter these emails.') %>
 <%= @unsubscribe_url %>
 
+<%= _('Disable Email Alerts') %>
```
Comment (lizconlan) [inline app/views/track_mailer/event_digest.text.erb]: Maybe we should call this "Unsubscribe From Email Alerts"? I'm not entirely sure on this but it is the language people seem to use/look for when they're trying to make the emails go away.

Or maybe make a ticket to review the wording in general 🤔

Comment (garethrees) [reply app/views/track_mailer/event_digest.text.erb]: I did think about this, but I initially thought that "Disable Email Alerts" makes a better distinction between the two actions. On reflection I think you're right – Unsubscribe is what you look for, and is clear enough. Fixed in cdf9513d844afed5b5d97c3bba4c3f1935441293.

File: app/views/users/email_alerts/destroy.html.erb (hunk 1 above)
Comment (garethrees) [reply app/views/users/email_alerts/destroy.html.erb]: Ah yes! Fixed in 220e08c8c.

I did want to take people directly to the wall, but the route requires a `url_name` param, so doing the redirect via signin gets really messy.

I did just try to add a route for `/profile/wall` to go to the current user's wall, but the controller doesn't handle that without modification, so will have to wait until another day.

File: app/views/track_mailer/event_digest.text.erb (hunk 2 above)
Comment (lizconlan) [reply app/views/track_mailer/event_digest.text.erb]: I also wasn't sure about this. The intent of disable is clearer but if I'm in a rush and become sufficiently irritated by something I don't remember signing up for to be considering the "Mark as Spam" button, I'll probably be skim-reading looking for "Unsubscribe"

**Pull request merged**